Added Git Configuration docs (#264)
epinio/docs: docs/references/git_configuration.md

---
sidebar_label: "Git Configuration"
title: ""
---

# Git Configuration

> Since version **1.10.0**, Epinio supports Git configuration.

Reviewer: ```suggestion
Starting with version **1.10.0**, Epinio supports Git configurations.
```
```suggestion
Epinio supports Git configurations since version **1.10.0**.
```

> You can provide for each git host particular settings, to clone private repositories, skip the SSL verification, or provide your own certificate bundle.

Reviewer: ```suggestion
Configurations provide information enabling the cloning of private repositories, disabling of SSL verification, or extending verification through a custom bundle of certificates. All on a per git host (+user/org, +repository) basis.
```

A Git configuration is a Kubernetes secret with the `epinio.io/api-git-credentials: "true"` label.

The available fields are:

- **url (required):** the host of the git instance
- **provider:** one of `github`, `gitlab`, `git`, `github_enterprise`, `gitlab_enterprise`
- **username:** used during the Basic Authentication 
- **password:** used during the Basic Authentication
- **userOrg:** used to restrict the configuration to a specific organization/project 
- **repo:** used to restrict the configuration to a specific repository 
- **skipSSL:** used to skip the SSL verification 
- **certificate:** the CA bundle to load for the SSL verification with self-signed certificates

> When importing a git repository Epinio will try to look for the existance of the most specific configuration. For example trying to clone the `https://github.com/myusername/myrepo` Epinio will first look for a configuration having the `https://github.com` *URL*, a `myusername` *userOrg* and a `myrepo` *repo*.

Reviewer: ```suggestion
When importing from a git repository Epinio will look for the the most specific matching configuration, if any. For example when trying to clone `https://github.com/myusername/myrepo` Epinio will first look for a configuration having the `https://github.com` *URL*, a `myusername` *userOrg* and a `myrepo` *repo*.
```

> If not found it will look for a configuration having the `https://github.com` *URL* and a `myusername` *userOrg*. And finally if not found Epinio will look for a configuration for the `https://github.com` *URL*.

Reviewer: ```suggestion
If not found it will look for a configuration having the `https://github.com` *URL* and a `myusername` *userOrg*. And finally it will look for a configuration having just the `https://github.com` *URL*.
```

> 

Reviewer 2: What happens if no matching configuration is found? Error returned I guess? Does it need to be stated explicitly or will it be clearly understood by the audience?

Reviewer: That looks fine to me. @jhkrug ?

Reviewer 2: Yep!
